### PR TITLE
Add packages and populate with module names

### PR DESCRIPTION
**Pull request by tlegen-k**
## Related Issues and Dependencies
https://github.com/thoth-station/datasets/pull/36 
…

## This introduces a breaking change

- [ ] Yes
- [x] No

## This should yield a new module release

- [ ] Yes
- [x] No

## This Pull Request implements

Add module names without namespaced calls from results to packages list to sync with database.

### CODE REVIEW

**Comment by tlegen-k** [comment]
Hi @pacospace 
Would you be able to take a look?
Next step will be to create new table linking PythonPackageVersion Table with the new table.

File: thoth/storages/graph/postgres.py
Comment (fridex) [inline thoth/storages/graph/postgres.py:5496]:
```suggestion
                    and package_call[0] != "_":
```

Comment (fridex) [reply thoth/storages/graph/postgres.py:5496]: Could you give examples for entries that are skipped?

**Review by fridex — COMMENTED** [review]
Looking at this, I'm not sure if we do not break `get_or_create` logic and create redundant entries (one with packages and the other without).

File: thoth/storages/graph/postgres.py
Comment (tlegen-k) [reply thoth/storages/graph/postgres.py]: > Could you give examples for entries that are skipped?

Hi @fridex, thank you for helping!
Yes, when exploring [Thoth solver dataset v2 ](https://www.kaggle.com/thothstation/thoth-solver-dataset-v10?select=thoth-solver-dataset-v2.0) for report with documet_id = solver-rhel-8-py38-210713042150-50168537c03f93c5 tree:packages looks like:
```
['_plotly_future_',
 '_plotly_utils',
 'plotly',
 'plotlywidget',
 '_plotly_utils.colors',
 'plotly.colors',
 'plotly.data',
 'plotly.express',
 'plotly.figure_factory',
 'plotly.graph_objs',
 'plotly.io',
 'plotly.matplotlylib',
 'plotly.offline',
 'plotly.plotly',
 'plotly.validators',
 'plotly.graph_objs.area',
...
]
```

Here we would like to skip all hidden entry imports that start with '_'.
Also AFAIK it is the only one that has two modules, but `plotlywidget` is not on pypi.org.

File: thoth/storages/graph/postgres.py
Comment (pacospace) [inline thoth/storages/graph/postgres.py:4121]: mandatory parameter cannot be after optional one :) is it optional? then I would put `Optional[str] = None`

File: thoth/storages/graph/models.py
Comment (pacospace) [inline thoth/storages/graph/models.py:75]: nullable should be True, initially the database won't have these entries and we cannot assign a default in this case, wdyt?

Comment (tlegen-k) [reply thoth/storages/graph/models.py]: > nullable should be True, initially the database won't have these entries and we cannot assign a default in this case, wdyt?

I agree, corrected in this commit d1a6c6a850b5e8f3e2577648f8c348225f492f8c.
Thanks for pointing!

File: thoth/storages/graph/postgres.py
Comment (tlegen-k) [reply thoth/storages/graph/postgres.py]: thank you)
I think it should be mandatory in this context 
corrected with c7996586173df63d59a6d46ac5a5063bc89bca20

**Comment by tlegen-k** [comment]
Hi @pacospace and @fridex,

Regards possible number of modules in tree:packages, apparently there are more cases when there is more than 1 module.
As an example this [notebook](https://colab.research.google.com/drive/1jF84zOzVTd5ThW5kvSE7C8fz4M9xYuwf?usp=sharing).

To give part of nonunique imports:
```
[['plotly', 'plotlywidget'],
 ['chart_studio', 'plotly', 'plotlywidget'],
 ['Cython', 'pyximport'],
 ['pkg_resources', 'setuptools'],
 ['plotly', 'plotlywidget'],
 ['pkg_resources', 'setuptools'],
 ['plotly', 'plotlywidget'],
 ['pkg_resources', 'setuptools'],
 ['plotly', 'plotlywidget'],
 ['pkg_resources', 'setuptools'],
 ['Cython', 'pyximport'],
 ['pkg_resources', 'setuptools'],
 ['Cython', 'pyximport'],
 ['plotly', 'plotlywidget'],
 ['pkg_resources', 'setuptools'],
 ['pkg_resources', 'setuptools'],
 ['caffe2', 'torch'],
...
```
They are under `nonunique_import_package` variable in the bottom of shared notebook.

**Comment by tlegen-k** [comment]
gently ping @fridex. Would you be able to give a feedback?

**Comment by fridex** [comment]
> gently ping @fridex. Would you be able to give a feedback?

As stated above, not sure if this would be a good solution. `get_or_create` logic on models creates a new entry in tables if there is no matching entry with all the attributes. Your code extends `PythonPackageVersion` which is used in other places of the project - see the adjusted part for sync_adviser_result as an example. This would mean that we eventually lose linkage to entries in the database if `import_package` will be set in some cases and in some cases it will be set to None (duplicate entries might be created).

**Comment by tlegen-k** [comment]
Hi @fridex,
Here is another attempt in doing correctly.
I have created `ImportPackages` class to hold only `document_id` and `import_packages` and extended logic of `sync_solver_result` to correctly initialize instance of `ImportPackages`.

Are there any other data fields that might be beneficial to store in `ImportPackages`?

**Comment by pacospace** [comment]
> > gently ping @fridex. Would you be able to give a feedback?
> 
> As stated above, not sure if this would be a good solution. `get_or_create` logic on models creates a new entry in tables if there is no matching entry with all the attributes. Your code extends `PythonPackageVersion` which is used in other places of the project - see the adjusted part for sync_adviser_result as an example. This would mean that we eventually lose linkage to entries in the database if `import_package` will be set in some cases and in some cases it will be set to None (duplicate entries might be created).

Should we make that mandatory? and resync all solvers with solver version that has that feature?

File: thoth/storages/graph/models.py
Comment (fridex) [inline thoth/storages/graph/models.py:1695]: This can be obtained from `PythonPackageVersion` directly so no need to keep it here.

```suggestion
```

File: thoth/storages/graph/models.py
Comment (fridex) [inline thoth/storages/graph/models.py:77]: You will need to add a relation to `ImportPackage`. I think you will need an association table for N-N relation here (as one package might provide the same import as another one):

https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#many-to-many

File: thoth/storages/graph/postgres.py
Comment (fridex) [inline thoth/storages/graph/postgres.py:5496]: No need to have any filtering here on names. You can directly create `ImportPackage` per each entry in `document["result"]["tree"]["packages"]`.

File: thoth/storages/graph/models.py
Comment (fridex) [inline thoth/storages/graph/models.py:1690]:
```suggestion
class ImportPackage(Base, BaseExtension):
```

File: thoth/storages/graph/models.py
Comment (fridex) [inline thoth/storages/graph/models.py:1697]:
```suggestion
    import_package = Column(Text, nullable=False)
```

File: thoth/storages/graph/models.py
Comment (tlegen-k) [reply thoth/storages/graph/models.py:83]: resolved by 8a428279967604e0e603a2556370045264245cbd

File: thoth/storages/graph/postgres.py
Comment (tlegen-k) [reply thoth/storages/graph/postgres.py]: resolved by 8a428279967604e0e603a2556370045264245cbd

**Comment by fridex** [comment]
> > > gently ping @fridex. Would you be able to give a feedback?
> > 
> > 
> > As stated above, not sure if this would be a good solution. `get_or_create` logic on models creates a new entry in tables if there is no matching entry with all the attributes. Your code extends `PythonPackageVersion` which is used in other places of the project - see the adjusted part for sync_adviser_result as an example. This would mean that we eventually lose linkage to entries in the database if `import_package` will be set in some cases and in some cases it will be set to None (duplicate entries might be created).
> 
> Should we make that mandatory? and resync all solvers with solver version that has that feature, so we don't create double entries? @fridex @tlegen-k

Making the field mandatory will not work - this information is present only in solver results and other sync methods also create `PythonPackageVersion` entries that would require a linkage to the newly introduced table.

File: thoth/storages/graph/postgres.py
Comment (pacospace) [inline thoth/storages/graph/postgres.py:5481]: this should run below https://github.com/thoth-station/storages/blob/e6fbbc34f782b8555729cf9f8fe270b2753f341f/thoth/storages/graph/postgres.py#L5332 :)

File: thoth/storages/graph/models.py
Comment (pacospace) [inline thoth/storages/graph/models.py:54]: let's follow some example we have in the databse like: https://github.com/thoth-station/storages/blob/e6fbbc34f782b8555729cf9f8fe270b2753f341f/thoth/storages/graph/models.py#L355

**Review by pacospace — COMMENTED** [review]
Once changes are done, new alembic migration to be added to the PR :)

File: thoth/storages/graph/postgres.py
Comment (tlegen-k) [reply thoth/storages/graph/postgres.py]: resolved by 8e65f38da030758bc68904136a0c0888e7ed54aa

File: thoth/storages/graph/postgres.py
Comment (pacospace) [inline thoth/storages/graph/postgres.py:319]: why this has changed?

File: thoth/storages/graph/models.py
Comment (pacospace) [inline thoth/storages/graph/models.py:53]: this should be a class as all tables

File: thoth/storages/graph/postgres.py
Comment (tlegen-k) [reply thoth/storages/graph/postgres.py:319]: To be able to run locally on macOS. Will make as it was before sending to prod.

File: thoth/storages/graph/models.py
Comment (tlegen-k) [reply thoth/storages/graph/models.py]: logic of connection between `ImportPackage` and `PPV` was changed to Association Object as requested by using `FoundImportPackage` in commit b4147fd5152048337263037bd70f41c6a96026af

**Comment by tlegen-k** [comment]
Hi @pacospace, thanks for continuous input!
I rewrote the model using association object instead of association table in b4147fd5152048337263037bd70f41c6a96026af.
Also created function `get_python_package_version_import_packages_all` that returns `ImportPackage` rows for the given `import_name`.
Next will be:
- [ ] Extend function so that: it outputs the PPV entries that are connected with the filtered 'ImportPackage`
- [ ] Create final alembic migration
Let me know what you think.

**Comment by fridex** [comment]
> Hi @pacospace, thanks for continuous input!
> I rewrote the model using association object instead of association table in [b4147fd](https://github.com/thoth-station/storages/commit/b4147fd5152048337263037bd70f41c6a96026af).
> Also created function `get_python_package_version_import_packages_all` that returns `ImportPackage` rows for the given `import_name`.
> Next will be:
> 
> * [ ]  Extend function so that: it outputs the PPV entries that are connected with the filtered 'ImportPackage`
> * [ ]  Create final alembic migration
>   Let me know what you think.

Might be a good idea to also include an index on `import_package_name` given the expected size of the newly created table. 👍🏻

**Comment by tlegen-k** [comment]
> > Hi @pacospace, thanks for continuous input!
> > I rewrote the model using association object instead of association table in [b4147fd](https://github.com/thoth-station/storages/commit/b4147fd5152048337263037bd70f41c6a96026af).
> > Also created function `get_python_package_version_import_packages_all` that returns `ImportPackage` rows for the given `import_name`.
> > Next will be:
> > 
> > * [ ]   Extend function so that: it outputs the PPV entries that are connected with the filtered 'ImportPackage`
> > * [ ]   Create final alembic migration
> >   Let me know what you think.
> 
> Might be a good idea to also include an index on `import_package_name` given the expected size of the newly created table. 👍🏻

Hi @pacospace and @fridex 
Added indexing in  739304e46a47c62c3c1e82c32383e086861525f8.
Final versions of `models.py` and `postgres.py`. Tested everything locally. Gives correct output.
For command `PYTHONPATH=. pipenv run python3 test_query.py` with following `test_query.py`:
```
from thoth.storages import GraphDatabase
from thoth.storages.sync import sync_solver_documents

# query part
graph = GraphDatabase()
graph.connect()


print(graph.get_python_package_version_import_packages_all("selinon"))
print(graph.get_python_package_version_import_packages_all("sklearn"))
print(graph.get_python_package_version_import_packages_all("dotenv"))
print(graph.get_python_package_version_import_packages_all("IPython"))
```
Output is: 
```
['selinon']
['scikit-learn']
['python-dotenv']
['ipython']
```
Which correctly identifies package name required to be installed.

File: thoth/storages/data/alembic/versions/c1839dfcdf13_add_importpackage_foundimportpackage_.py
Comment (pacospace) [inline thoth/storages/data/alembic/versions/c1839dfcdf13_add_importpackage_foundimportpackage_.py:33]: please remove line 33 to 40 from the migration manually as they are not related to this. cc @KPostOffice that encountered similar problem.

File: thoth/storages/graph/postgres.py
Comment (pacospace) [inline thoth/storages/graph/postgres.py:6794]: return type should be `List[str]`

**Review by pacospace — COMMENTED** [review]
Few comments and please put back parts that you commented/modified for testing. Very nice work 👏🏼

File: thoth/storages/data/alembic/versions/c1839dfcdf13_add_importpackage_foundimportpackage_.py
Comment (tlegen-k) [reply thoth/storages/data/alembic/versions/c1839dfcdf13_add_importpackage_foundimportpackage_.py]: removed in 152184ab2cafbc52b8c8171b91a5f8317e02d60e

File: thoth/storages/graph/postgres.py
Comment (tlegen-k) [reply thoth/storages/graph/postgres.py]: fixed in 3f36204cb14d6f6cd76d9bdb7357c33dcfa5ad51

File: thoth/storages/data/alembic/versions/c1839dfcdf13_add_importpackage_foundimportpackage_.py
Comment (pacospace) [inline thoth/storages/data/alembic/versions/c1839dfcdf13_add_importpackage_foundimportpackage_.py:39]: Also line 39 to 44 should be manually removed

Comment (tlegen-k) [reply thoth/storages/data/alembic/versions/c1839dfcdf13_add_importpackage_foundimportpackage_.py]: was removed

**Comment by tlegen-k** [comment]
> Few comments and please put back parts that you commented/modified for testing. Very nice work 👏🏼

Thank you for review @pacospace!
I made sure to return part that were commented to run locally.
What should I work on next?

File: thoth/storages/graph/models.py
Comment (pacospace) [inline thoth/storages/graph/models.py:1699]: I actually don't see the index created for `import_package_name` as @fridex suggested. It's missing the migration also indeed, there should be something like `op.create_index`

Comment (tlegen-k) [reply thoth/storages/graph/models.py:1699]: > I actually don't see the index created for `import_package_name` as @fridex suggested. It's missing the migration also indeed, there should be something like `op.create_index`

I understood as `id` field is index in line 1695:
`id = Column(Integer, primary_key=True, autoincrement=True, nullable=False)`

Screen Shot 2021-08-12 at 5 33 48 PM

Or if it is not that i can definitely work on that.

**Comment by goern** [comment]
@all what is the status of this one? does it need a little love before ready to merge?

File: thoth/storages/graph/models.py
Comment (pacospace) [reply thoth/storages/graph/models.py:1699]: See for example: https://github.com/thoth-station/storages/blob/61b9de2fa41a31547bb1b45ce7b21276455e70e7/thoth/storages/graph/models.py#L215. I think @fridex meant to have one specific for that, the ID has an index by default I think.

Comment (tlegen-k) [reply thoth/storages/graph/models.py:1699]: Now clear. Implemented in d8d35af04d26d99a2f8187c78f267a6639e026a6
And created new alembic migration with adjustments to it that you pointed out previously

Comment (tlegen-k) [reply thoth/storages/graph/models.py:1699]: > See for example:
> 
> https://github.com/thoth-station/storages/blob/61b9de2fa41a31547bb1b45ce7b21276455e70e7/thoth/storages/graph/models.py#L215
> . I think @fridex meant to have one specific for that, the ID has an index by default I think.

Now clear. Implemented in d8d35af04d26d99a2f8187c78f267a6639e026a6
And created new alembic migration with adjustments to it that you pointed out previously
Let me know if there is anything else)

File: thoth/storages/graph/postgres.py
Comment (fridex) [inline thoth/storages/graph/postgres.py:5504]:
```suggestion
```

File: thoth/storages/graph/postgres.py
Comment (fridex) [inline thoth/storages/graph/postgres.py:5466]:
```suggestion
```

**Review by fridex — COMMENTED** [review]
Nice! The code looks good to me (haven't tried it thought). Could you please squash commits to keep the git history clean? It might help in the future with debugging.

Nice work! 💯 👍🏻

**Comment by tlegen-k** [comment]
> Nice! The code looks good to me (haven't tried it thought). Could you please squash commits to keep the git history clean? It might help in the future with debugging.
> 
> Nice work! 💯 👍🏻

Thank you @fridex! Have squashed 26 commits using git rebase.

**Comment by pacospace** [comment]
> > Nice! The code looks good to me (haven't tried it thought). Could you please squash commits to keep the git history clean? It might help in the future with debugging.
> > Nice work! 💯 👍🏻
> 
> Thank you @fridex! Have squashed 26 commits using git rebase.

Can you please adjust to make pre-commit happy? :)

**Comment by tlegen-k** [comment]
> > > Nice! The code looks good to me (haven't tried it thought). Could you please squash commits to keep the git history clean? It might help in the future with debugging.
> > > Nice work! 💯 👍🏻
> > 
> > 
> > Thank you @fridex! Have squashed 26 commits using git rebase.
> 
> Can you please adjust to make pre-commit happy? :)

Totally overlooked that one)
Thanks for pointing out @pacospace 
Fixed now.  Only needs approval labels.

**Comment by pacospace** [comment]
> > > > Nice! The code looks good to me (haven't tried it thought). Could you please squash commits to keep the git history clean? It might help in the future with debugging.
> > > > Nice work! 💯 👍🏻
> > > 
> > > 
> > > Thank you @fridex! Have squashed 26 commits using git rebase.
> > 
> > 
> > Can you please adjust to make pre-commit happy? :)
> 
> Totally overlooked that one)
> Thanks for pointing out @pacospace
> Fixed now. Only needs approval labels.

Thanks!! We just miss rebase after https://github.com/thoth-station/storages/pull/2384#issuecomment-901445109 was merged and please adjust the alembic migration number now, because that PR already introduced a migration :)

File: thoth/storages/graph/postgres.py
Comment (fridex) [inline thoth/storages/graph/postgres.py:6823]: Can the query return `package_name`, `package_version`, `index_url` triplet so the package is uniquely identified?

**Review by fridex — APPROVED** [review]
lgtm, thanks @tlegen-k 💯 👍🏻

**Comment by tlegen-k** [comment]
> lgtm, thanks @tlegen-k 💯 👍🏻

Thank you for helping!

**Pull request merged**
